File: cpp/1011_在D天内送达包裹的能力_medium.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int D) {
        int left = *max_element(weights.begin(), weights.end());
        int right = accumulate(weights.begin(), weights.end(), 0);
        while (left < right) {
            int mid = (left + right) >> 1;
            int need = 1, cur = 0;
            for (auto w : weights) {
                if (cur + w > mid) {
                    cur = 0;
                    need++;
                    if (need > D) break;
                }
                cur += w;
            }
            if (need > D) left = mid + 1;
            else right = mid;
        }
        return left;
    }
};

File: python/1011_在D天内送达包裹的能力_medium.py
class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        left, right = max(weights), sum(weights)
        while left < right:
            mid = (left + right) >> 1
            need, cur = 1, 0
            for w in weights:
                if cur + w > mid:
                    cur = 0
                    need += 1
                    if need > D:
                        break
                cur += w
            if need > D:
                left = mid + 1
            else:
                right = mid
        return left
